### ISSUE TITLE
How to connect to EXASOL with encryption turned on

### ISSUE THREAD

**Issue by GunayAnach** [EXASOL/websocket-api#7]
I am looking to connect to exasol which has encryption turned on through python. It connects without issues to non encrypted database.

I tried to pass the _ENCRYPTION=Y_ as described in the manual but no success. 

> ENCRYPTION Switches on the automatic encryption. Valid values are "Y" and "N" (default
is "N").

Seems like it ignores it.
Keeps blasting - **OperationalError: database error [08004] Connection exception - Client connection must be encrypted.**
```
database_table = "SCHEMA_NAME.MY_TABLE"
conn = EXASOL.connect('ws://192.168.56.101:8563', 
                      'UID=sys', 
                      'PWD=pwd', 
                      'ENCRYPTION=Y', 
                      )

sql = "SELECT * FROM %s LIMIT 1000;" % database_table
df_sql = pd.read_sql(sql, conn)

```

```

/home/ec2-user/anaconda3/envs/mxnet_p27/lib/python2.7/site-packages/IPython/core/interactiveshell.pyc in run_cell_magic(self, magic_name, line, cell)
   2115             magic_arg_s = self.var_expand(line, stack_depth)
   2116             with self.builtin_trap:
-> 2117                 result = fn(magic_arg_s, cell)
   2118             return result
   2119 

<decorator-gen-60> in time(self, line, cell, local_ns)

/home/ec2-user/anaconda3/envs/mxnet_p27/lib/python2.7/site-packages/IPython/core/magic.pyc in <lambda>(f, *a, **k)
    186     # but it's overkill for just that one bit of state.
    187     def magic_deco(arg):
--> 188         call = lambda f, *a, **k: f(*a, **k)
    189 
    190         if callable(arg):

/home/ec2-user/anaconda3/envs/mxnet_p27/lib/python2.7/site-packages/IPython/core/magics/execution.pyc in time(self, line, cell, local_ns)
   1183         else:
   1184             st = clock2()
-> 1185             exec(code, glob, local_ns)
   1186             end = clock2()
   1187             out = None

<timed exec> in <module>()

/home/ec2-user/anaconda3/envs/mxnet_p27/lib/python2.7/site-packages/EXASOL/__init__.pyc in __init__(self, url, username, password, autocommit, queryTimeout, useCompression, typeMapper)
    324         self._inconnect = False
    325         self._type_mapper = typeMapper
--> 326         self._connect()
    327         self._login()
    328 

/home/ec2-user/anaconda3/envs/mxnet_p27/lib/python2.7/site-packages/EXASOL/__init__.pyc in _connect(self)
    417                 self._ws_send = self.__ws.send
    418                 self._ws_recv = self.__ws.recv
--> 419                 ret = self._req(command = 'login', protocolVersion = 1)
    420                 if CRYPTO_LIB == 'rsa':
    421                     if sys.version_info.major >= 3:

/home/ec2-user/anaconda3/envs/mxnet_p27/lib/python2.7/site-packages/EXASOL/__init__.pyc in _req(self, **req)
    407             raise OperationalError('%s [%s] %s' % ('database error',
    408                                                    rep['exception']['sqlCode'],
--> 409                                                    rep['exception']['text']))
    410         raise OperationalError('internal error: %s' % repr(rep))
    411 

OperationalError: database error [08004] Connection exception - Client connection must be encrypted.

```

**Comment by exazg** [comment]
Please have a look at the usage example at https://github.com/EXASOL/websocket-api/tree/master/python which shows how to create a connection. Also note that all WebSocket-related documentation is here on GitHub. Other interfaces described in the manual (e.g. ODBC) are not relevant to WebSocket connections.

Unencrypted WebSocket connections use the ```ws``` URI scheme while encrypted WebSocket connections (WebSocket Secure) use the ```wss``` scheme (similar to ```http``` and ```https```). See https://en.wikipedia.org/wiki/WebSocket for details.

So your connect() call should look like the following.
```
with EXASOL.connect('wss://192.168.56.101:8563', 'sys', 'pwd') as connection:
```
Please let us know if that fixes your problem or if there are any further issues.

**Comment by GunayAnach** [comment]
Thanks for the speedy reply @exazg, seems like it connects with _wss_, but this time it fails with 
**SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)**

Not sure why it needs certificate to query exasol, there is no certification setup between two ec2 systems as I know of.

My setup is: Exasol is installed on AWS and I am trying to pull some data into AWS Sagemaker jupyter instance running python 2.7. Thats why ODBC setup doesn't quite work for me and i need to rely on websockets. Any other suggested methods would be appreciated.

**Comment by exazg** [comment]
Currently there's not an option to disable SSL certificate verification using the EXASOL.connect() method. However, you can quickly disable it manually--at least for testing purposes--by making the two small changes in the diff below to EXASOL/\_\_init\_\_.py (https://github.com/EXASOL/websocket-api/blob/master/python/EXASOL/__init__.py).

```
--- __init__.py.orig	2018-04-05 17:18:54.269481694 +0200
+++ __init__.py	2018-04-05 17:20:49.219235989 +0200
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 
-import sys, datetime, platform, getpass, base64, time, zlib
+import sys, datetime, platform, getpass, base64, time, zlib, ssl
 from websocket import create_connection
 from pprint import pprint as pp
 from json import loads, dumps
@@ -413,7 +413,7 @@
         with timer(self, 'conn'):
             try:
                 self._inconnect = True
-                self.__ws = create_connection(self._url)
+                self.__ws = create_connection(self._url, sslopt={"cert_reqs": ssl.CERT_NONE})
                 self._ws_send = self.__ws.send
                 self._ws_recv = self.__ws.recv
                 ret = self._req(command = 'login', protocolVersion = 1)
```

Hopefully this will get you going. We'll consider allowing such options to be simply specified in EXASOL.connect() in the future as well.

**Comment by GunayAnach** [comment]
@exazg you are star!
Got it working, if anyone needs it disabled until such functionality is implemented can get it from [here](https://github.com/GunayAnach/websocket-api).

**Comment by exazg** [comment]
I just fixed this issue in #8 and updated the documentation as well. Please have a look at it and let us know if you have any issues. Thanks again for reporting this!
